### PR TITLE
KUBOS-107 - HAL Testing 

### PR DESCRIPTION
**Pull request by rplauche**
Note - This PR is currently for in-progress reviewing. A future comment and assigning of reviewers will indicate initial code completeness.

Goals:
* Create tests to exercise HAL interface functionality
   * UART
   * I2C
   * SPI
* Native (linux) test compatibility for quick building/verification
   * Native testing depends on changes to targets.....PRs pending
* Tests should also run and pass on target
* Tests will be created in the yotta test folder

### CODE REVIEW

File: test/i2c.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+static void test_init_write(void)
+{
+    char data = 'A';
+    KI2CConf conf = get_conf();
+
+    k_i2c_init(TEST_I2C, &conf);
+
+    TEST_ASSERT_EQUAL_INT(k_i2c_write(TEST_I2C, TEST_ADDR, &data, 1), I2C_OK);
+
+    k_i2c_terminate(TEST_I2C);
```
Comment (CatFreed) [inline test/i2c.c]: If the TEST_ASSERT fails, terminate won't be called.  Might be prudent to terminate before the write result is checked.

File: test/i2c.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+static void test_init_term_write(void)
+{
+    KI2CConf conf = get_conf();
+
+    k_i2c_init(TEST_I2C, &conf);
+
+    k_i2c_terminate(TEST_I2C);
+
+    char data = 'A';
```
Comment (CatFreed) [inline test/i2c.c]: The other tests have this variable declaration at the beginning

File: test/i2c.c [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    RUN_TEST(test_no_init_read);
+    RUN_TEST(test_init_write);
+    RUN_TEST(test_init_write_null);
+    RUN_TEST(test_init_read);
+    RUN_TEST(test_init_read_null);
+    RUN_TEST(test_init_write_read);
+    RUN_TEST(test_init_term_write);
+    RUN_TEST(test_init_term_read);
+    RUN_TEST(test_init_term_write_read);
+    RUN_TEST(test_init_term_init_write_read);
```
Comment (CatFreed) [inline test/i2c.c]: Possibly add test cases where the write/read length is longer than the variable to be written/read?

File: test/i2c.c (hunk 1 above)
Comment (rplauche) [inline test/i2c.c]: Ahh yes that is a good catch.

File: test/i2c.c (hunk 3 above)
Comment (rplauche) [inline test/i2c.c]: I've added that test case for the uart interface but not for i2c and spi. I think for those interfaces that test case would make more sense at the lower hal level, as the functions at this level are just pass throughs.

**Comment by CatFreed** [comment]
Looks good to me

File: source/i2c.c [hunk 4]
```diff
@@ -80,21 +85,21 @@ KI2CStatus k_i2c_read(KI2CNum i2c, uint16_t addr, uint8_t* ptr, int len)
...
-            xSemaphoreGive(ki2c->i2c_lock);
+            csp_mutex_unlock(&(ki2c->i2c_lock));
         }
     }
     return ret;
 }
 
 KI2C* kprv_i2c_get(KI2CNum i2c)
 {
-    return &k_i2cs[i2c];
```
Comment (rplauche) [inline source/i2c.c]: Need to do this for `kprv_spi_get` as well

**Pull request merged**